### ISSUE TITLE
lxc snapshot does not run properly under kernel version 5.10.2 but it works as expected under 5.4.83

### ISSUE THREAD

**Issue by graysky2** [lxc/lxc#3609]
# Required information

 * Distribution: Both ArchARM (aarch64) and also Arch Linux (x86_64).  
 * **Note - I do not believe this is a problem unique to ArchARM since I get a similar problem on Arch Linux on Intel x86_64 as well.  For the sake of simplicity, the rest of this ticket shows data from ArchARM since I experienced the bug there first.**
 * Distribution version: Rolling release
 * The output of
   * ```lxc-start --version 4.0.5```
   *
```
lxc-checkconfig
LXC version 4.0.5
--- Namespaces ---
Namespaces: enabled
Utsname namespace: enabled
Ipc namespace: enabled
Pid namespace: enabled
User namespace: enabled
Warning: newuidmap is not setuid-root
Warning: newgidmap is not setuid-root
Network namespace: enabled

--- Control groups ---
Cgroups: enabled

Cgroup v1 mount points: 
/sys/fs/cgroup/systemd
/sys/fs/cgroup/freezer
/sys/fs/cgroup/cpu,cpuacct
/sys/fs/cgroup/perf_event
/sys/fs/cgroup/pids
/sys/fs/cgroup/blkio
/sys/fs/cgroup/cpuset
/sys/fs/cgroup/net_cls,net_prio
/sys/fs/cgroup/devices

Cgroup v2 mount points: 
/sys/fs/cgroup/unified

Cgroup v1 clone_children flag: enabled
Cgroup device: enabled
Cgroup sched: enabled
Cgroup cpu account: enabled
Cgroup memory controller: enabled
Cgroup cpuset: enabled

--- Misc ---
Veth pair device: enabled, not loaded
Macvlan: enabled, not loaded
Vlan: enabled, not loaded
Bridges: enabled, loaded
Advanced netfilter: enabled, not loaded
CONFIG_NF_NAT_IPV4: missing
CONFIG_NF_NAT_IPV6: missing
CONFIG_IP_NF_TARGET_MASQUERADE: enabled, not loaded
CONFIG_IP6_NF_TARGET_MASQUERADE: enabled, not loaded
CONFIG_NETFILTER_XT_TARGET_CHECKSUM: enabled, loaded
CONFIG_NETFILTER_XT_MATCH_COMMENT: enabled, loaded
FUSE (for use with lxcfs): enabled, loaded

--- Checkpoint/Restore ---
checkpoint restore: enabled
CONFIG_FHANDLE: enabled
CONFIG_EVENTFD: enabled
CONFIG_EPOLL: enabled
CONFIG_UNIX_DIAG: missing
CONFIG_INET_DIAG: enabled
CONFIG_PACKET_DIAG: enabled
CONFIG_NETLINK_DIAG: missing
File capabilities: 
```
   * `uname -a`
  
```
Linux enigma64 5.10.2-2-ARCH #1 SMP PREEMPT Wed Dec 23 11:16:14 EST 2020 aarch64 GNU/Linux
```
   
* `cat /proc/self/cgroup`
 
```
9:devices:/user.slice
8:net_cls,net_prio:/
7:cpuset:/
6:blkio:/user.slice
5:pids:/user.slice/user-1000.slice/session-1.scope
4:perf_event:/
3:cpu,cpuacct:/user.slice
2:freezer:/
1:name=systemd:/user.slice/user-1000.slice/session-1.scope
0::/user.slice/user-1000.slice/session-1.scope
```
   * `cat /proc/1/mounts`
```
/dev/root / ext4 rw,relatime 0 0
devtmpfs /dev devtmpfs rw,relatime,size=3966640k,nr_inodes=991660,mode=755 0 0
proc /proc proc rw,nosuid,nodev,noexec,relatime 0 0
sysfs /sys sysfs rw,nosuid,nodev,noexec,relatime 0 0
securityfs /sys/kernel/security securityfs rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /dev/shm tmpfs rw,nosuid,nodev 0 0
devpts /dev/pts devpts rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000 0 0
tmpfs /run tmpfs rw,nosuid,nodev,size=1600240k,nr_inodes=819200,mode=755 0 0
tmpfs /sys/fs/cgroup tmpfs ro,nosuid,nodev,noexec,size=4096k,nr_inodes=1024,mode=755 0 0
cgroup2 /sys/fs/cgroup/unified cgroup2 rw,nosuid,nodev,noexec,relatime,nsdelegate 0 0
cgroup /sys/fs/cgroup/systemd cgroup rw,nosuid,nodev,noexec,relatime,xattr,name=systemd 0 0
none /sys/fs/bpf bpf rw,nosuid,nodev,noexec,relatime,mode=700 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
cgroup /sys/fs/cgroup/cpu,cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpu,cpuacct 0 0
cgroup /sys/fs/cgroup/perf_event cgroup rw,nosuid,nodev,noexec,relatime,perf_event 0 0
cgroup /sys/fs/cgroup/pids cgroup rw,nosuid,nodev,noexec,relatime,pids 0 0
cgroup /sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
cgroup /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset 0 0
cgroup /sys/fs/cgroup/net_cls,net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_cls,net_prio 0 0
cgroup /sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
systemd-1 /proc/sys/fs/binfmt_misc autofs rw,relatime,fd=30,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=10741 0 0
mqueue /dev/mqueue mqueue rw,nosuid,nodev,noexec,relatime 0 0
tracefs /sys/kernel/tracing tracefs rw,nosuid,nodev,noexec,relatime 0 0
debugfs /sys/kernel/debug debugfs rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /tmp tmpfs rw,nosuid,nodev,size=4000596k,nr_inodes=409600 0 0
fusectl /sys/fs/fuse/connections fusectl rw,nosuid,nodev,noexec,relatime 0 0
configfs /sys/kernel/config configfs rw,nosuid,nodev,noexec,relatime 0 0
/dev/mmcblk0p1 /boot vfat rw,relatime,fmask=0022,dmask=0022,codepage=437,iocharset=ascii,shortname=mixed,errors=remount-ro 0 0
tmpfs /run/user/1000 tmpfs rw,nosuid,nodev,relatime,size=800116k,nr_inodes=200029,mode=700,uid=1000,gid=985 0 0
```

# Issue description

Running a contain snapshot does not work properly once started under kernel version 5.10.2 but it works as expected under 5.4.83.

# Steps to reproduce

**Note 1 - I am running https://github.com/lxc/lxc/commit/6e8b98aaf89a99aa53b5a41f397741f066fbb704 from the stable-4.0 branch.**
**Note 2 - The parent container seems to work as expected, this issue only seems to affect the overlayfs snapshot of it.**

1. Boot into the 5.10.2 kernel and create a snapshot: `lxc-copy -n basedog -N test -M -s -B overlayfs`
2. Start the snapshot: `lxc-start -n test -F`  ... here startup hangs:
```
[ ***  ] A start job is running for Create Volatile Files and Directories (47s / no limit)
```

I added the following to `/var/lib/lxc/basedog/rootfs/usr/lib/systemd/system/systemd-tmpfiles-setup.service` under the `[Service]` section:
```
TimeoutStartSec=15
```

Upon starting the "test" container, I noticed systemd increased that 15s timeout several time up to 5m and it eventually worked.  I found this in my journalctl output:
```
Dec 27 12:11:44 enigma64 kernel: INFO: task systemd-tmpfile:778 blocked for more than 362 seconds.
Dec 27 12:11:44 enigma64 kernel:       Tainted: G         C        5.10.3-1-ARCH #1
Dec 27 12:11:44 enigma64 kernel: "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
Dec 27 12:11:44 enigma64 kernel: task:systemd-tmpfile state:D stack:    0 pid:  778 ppid:   703 flags:0x00000801
Dec 27 12:11:44 enigma64 kernel: Call trace:
Dec 27 12:11:44 enigma64 kernel:  __switch_to+0x9c/0xe0
Dec 27 12:11:44 enigma64 kernel:  __schedule+0x2a0/0x924
Dec 27 12:11:44 enigma64 kernel:  schedule+0x44/0xfc
Dec 27 12:11:44 enigma64 kernel:  rwsem_down_write_slowpath+0x238/0x660
Dec 27 12:11:44 enigma64 kernel:  down_write+0x5c/0x70
Dec 27 12:11:44 enigma64 kernel:  ovl_dir_real_file+0xd0/0x120 [overlay]
Dec 27 12:11:44 enigma64 kernel:  ovl_real_fdget+0x40/0x94 [overlay]
Dec 27 12:11:44 enigma64 kernel:  ovl_real_ioctl+0x3c/0xbc [overlay]
Dec 27 12:11:44 enigma64 kernel:  ovl_ioctl_set_flags+0x11c/0x174 [overlay]
Dec 27 12:11:44 enigma64 kernel:  ovl_ioctl+0xc4/0x214 [overlay]
Dec 27 12:11:44 enigma64 kernel:  __arm64_sys_ioctl+0x49c/0xc80
Dec 27 12:11:44 enigma64 kernel:  do_el0_svc+0x88/0x210
Dec 27 12:11:44 enigma64 kernel:  el0_svc+0x14/0x20
Dec 27 12:11:44 enigma64 kernel:  el0_sync_handler+0x1a4/0x1b0
Dec 27 12:11:44 enigma64 kernel:  el0_sync+0x174/0x180
```

It should be noted that although the copy finally started, it did not run as expected.  For example, it would not stop when called to do so.  I just hung up.

```
# lxc-stop -n test
```

And since I started it in foreground mode I found it just hung up on `Starting Halt...`

# Information to attach

 - [ ] any relevant kernel output (`dmesg`).. there is nothing in dmesg
 - [x] container log (The <log> file from running `lxc-start -n <c> -l TRACE -o <logfile> `)
 - [x] the containers configuration file

[basedog.txt](https://github.com/lxc/lxc/files/5744507/basedog.txt)
[test.txt](https://github.com/lxc/lxc/files/5744508/test.txt)
[container.log](https://github.com/lxc/lxc/files/5744511/container.log)

**Comment by graysky2** [comment]
I should add that if I do step 1 then reboot, then start the container, it boots as it should.  Very peculiar.

**Comment by brauner** [comment]
Smells like a kernel bug. I'll take a look now. Sorry for the delay.

**Comment by graysky2** [comment]
No need to apologize, I hope you had a restful and happy break!

**Comment by brauner** [comment]
Does arch already provide v5.11-rc1 somewhere?

**Comment by brauner** [comment]
This seems fixed with v5.11-rc1.

**Comment by brauner** [comment]
I'm closing this since it's a kernel regression and it seems that this is fixed on newer kernels. I'd wait for a new stable kernel update.

**Comment by graysky2** [comment]
Hopefully whatever broke will be backported and endup in 5.10.x series.  Thanks for demonstrating that you do not see this bug on 5.11-rc1.

**Comment by graysky2** [comment]
@brauner - I'd like to request that you re-open this task.  I am getting the same bug manifesting under 5.11-rc2.

**Comment by brauner** [comment]
Huh. Ok, I'll look into it soon!

**Comment by graysky2** [comment]
Just as an update, 5.10.7 gives the same behavior.  I have not tested 5.11-rc3.
```
[  OK  ] Started Network Service.
[  OK  ] Finished CLI Netfilter Manager.
[***   ] A start job is running for Create Volatile Files and Directories (51s / no limit)
```

**Comment by msilletti** [comment]
Yes it is still happening with 5.10.13-rc1 and 5.11-rc6.

**Comment by brauner** [comment]
Sorry, I haven't forgotten the issue, I promise. But this is likely a kernel bug and I'd need ~2 weeks and then I can spend some more time on this.

**Comment by graysky2** [comment]
@brauner - Thanks for your support with this.
@msilletti - Assuming you're using a systemd distro, if you wait longer on reboots, systemd eventually kills the hung processes without filesystem corruption.

**Comment by graysky2** [comment]
@brauner - I believe this is resolved under 5.10.16 and under 5.11.0.
